src/sentry/api/endpoints/auth_register.py

```diff
@@ -0,0 +1,140 @@
+import logging
+
+from django.conf import settings
+from django.core.exceptions import ValidationError as DjangoValidationError
+from django.db import transaction
+from rest_framework import serializers
+
+from sentry import newsletter, options
+from sentry.api.base import Endpoint
+from sentry.api.serializers.base import serialize
+from sentry.api.serializers.rest_framework.base import CamelSnakeSerializer
+from sentry.app import ratelimiter
+from sentry.auth import password_validation
+from sentry.models import User
+from sentry.signals import user_signup
+from sentry.utils import auth, metrics
+
+logger = logging.getLogger("sentry.auth")
+audit_logger = logging.getLogger("sentry.audit.ui")
+
+ERR_UNKNOWN = "An unknown error occurred while creating the account."
+
+
+def send_confirm_emails(user):
+    for email in user.emails.filter(is_verified=False):
+        user.send_confirm_email_singular(email=email, is_new_user=True)
+
+
+class AuthRegisterSerializer(CamelSnakeSerializer):
+    name = serializers.CharField(max_length=30, required=True)
+    email = serializers.EmailField(required=True)
+    password = serializers.CharField(required=True)
+    subscribe = serializers.BooleanField(required=False)
+
+    referrer = serializers.CharField(required=False)
+
+    def is_rate_limited(self):
+        limit = options.get("auth.ip-rate-limit")
+        if not limit:
+            return False
+
+        ip_address = self.context["request"].META["REMOTE_ADDR"]
+        return ratelimiter.is_limited(f"auth:ip:{ip_address}", limit)
+
+    def validate_email(self, value):
+        value = value.strip()
+        if not value:
+            return
+        if User.objects.filter(username__iexact=value).exists():
+            raise serializers.ValidationError(
+                "An account is already registered with that email address."
+            )
+        return value.lower()
+
+    def validate_password(self, value):
+        try:
+            password_validation.validate_password(value)
+        except DjangoValidationError as error:
+            raise serializers.ValidationError(error.messages[0])
+        return value
+
+    def validate(self, data):
+        if self.is_rate_limited():
+            raise serializers.ValidationError("Rate limit exceeded")
+        return data
+
+    def create(self, data) -> User:
+        request = self.context["request"]
+
+        user = User(username=data["email"], email=data["email"], name=data["name"])
+        user.set_password(data["password"])
+        user.save()
+
+        # Authenticate user into the current request context
+        #
+        # XXX: This is a side-effect, which is maybe a little confusing as
+        #      creation of an account using this serializer will also
+        #      authenticate the created user.
+        user.backend = settings.AUTHENTICATION_BACKENDS[0]
```

Comment:
realistically we probably shouldn't be doing this given an API can be queried many ways.. its not obvious to me if this creates an attack vector (I dont think it does), but it at the very least would create session constructs where none is needed. that'd likely require other changes on our frontend though so for now its ok to keep this.

Reply:
auth_login has same issue - we also probably could change this to simply call `auth.login()` though that has some complexity that shouldn't be very relevant here